File: WebAppFirewall/Classifier/Classifier/classifier.py
# Existing imports
import numpy as np
from collections import defaultdict
from scipy.spatial.distance import cdist
from scipy.sparse import csr_matrix

def create_float_defaultdict():
    """
    Creates a defaultdict with float as the default factory.
    
    Returns:
        defaultdict: A defaultdict that returns 0.0 for missing keys.
    """
    return defaultdict(float)

def create_int_defaultdict():
    """
    Creates a defaultdict with int as the default factory.
    
    Returns:
        defaultdict: A defaultdict that returns 0 for missing keys.
    """
    return defaultdict(int)

class MultinomialNaiveBayes:
    """
    Implements the Multinomial Naive Bayes algorithm for classification.
    """

    def __init__(self, alpha=1.0):
        """
        Initializes the MultinomialNaiveBayes classifier.

        Args:
            alpha (float): Smoothing parameter for Laplace smoothing. Default is 1.0.
        """
        self.alpha = alpha  # Smoothing parameter
        self.class_priors = {}
        self.feature_probs = defaultdict(create_float_defaultdict)
        self.classes = set()
        self.vocabulary = set()

    def fit(self, X, y):
        """
        Fits the Multinomial Naive Bayes classifier to the training data.

        Args:
            X (scipy.sparse.csr_matrix): The input features as a sparse matrix.
            y (array-like): The target labels.
        """
        n_samples, n_features = X.shape
        self.classes = set(y)
        
        # Calculate class priors
        class_counts = defaultdict(int)
        for label in y:
            class_counts[label] += 1
        
        for c in self.classes:
            self.class_priors[c] = class_counts[c] / n_samples

        # Calculate feature probabilities
        feature_counts = defaultdict(create_int_defaultdict)
        class_totals = defaultdict(int)

        for i in range(n_samples):
            c = y[i]
            for j in X[i].indices:
                feature_counts[c][j] += X[i, j]
                class_totals[c] += X[i, j]
                self.vocabulary.add(j)

        for c in self.classes:
            for feature in self.vocabulary:
                numerator = feature_counts[c][feature] + self.alpha
                denominator = class_totals[c] + self.alpha * len(self.vocabulary)
                self.feature_probs[c][feature] = numerator / denominator

    def predict(self, X):
        """
        Predicts the class labels for the input samples.

        Args:
            X (scipy.sparse.csr_matrix): The input features as a sparse matrix.

        Returns:
            list: Predicted class labels for each input sample.
        """
        return [self._predict_single(x) for x in X]

    def _predict_single(self, x):
        """
        Predicts the class label for a single input sample.

        Args:
            x (scipy.sparse.csr_matrix): A single input sample as a sparse matrix row.

        Returns:
            The predicted class label.
        """
        best_class = None
        best_score = float('-inf')

        for c in self.classes:
            score = np.log(self.class_priors[c])
            for i in x.indices:
                score += x[0, i] * np.log(self.feature_probs[c][i])
            
            if score > best_score:
                best_score = score
                best_class = c

        return best_class

class OneClassSVMClassifier:
    """
    One-Class SVM Classifier for outlier detection using various kernels.

    Parameters
    ----------
    kernel : str, optional (default='rbf')
        Specifies the kernel type to be used in the algorithm. Supported options are 'rbf''.
        
    nu : float, optional (default=0.5)
        An upper bound on the fraction of margin errors and a lower bound of the fraction of support vectors. Should be in the interval (0, 1].

    gamma : {'scale', float}, optional (default='scale')
        Kernel coefficient for 'rbf' and 'poly'. If 'scale', it uses 1 / (n_features * X.var()) as value of gamma.

    degree : int, optional (default=3)
        Degree of the polynomial kernel function ('poly'). Ignored by other kernels.

    coef0 : float, optional (default=1)
        Independent term in polynomial kernel. It is only significant in 'poly' kernel.

    Attributes
    ----------
    alpha : ndarray of shape (n_samples,)
        Lagrange multipliers for the support vectors.

    support_vectors : ndarray of shape (n_samples, n_features)
        Support vectors.

    intercept : float
        Intercept term in the decision function.

    Methods
    -------
    fit(data)
        Fit the model using the training data.

    predict(data)
        Perform classification on samples in data.
    """
    def __init__(self, kernel='rbf', nu=0.5, gamma='scale', degree=3, coef0=1):
        self.kernel = kernel
        self.nu = nu
        self.gamma = gamma
        self.degree = degree
        self.coef0 = coef0
        self.alpha = None
        self.support_vectors = None
        self.intercept = None

    def _rbf_kernel(self, X, Y):
        if self.gamma == 'scale':
            self.gamma = 1 / (X.shape[1] * X.var())
        return np.exp(-self.gamma * cdist(X, Y, 'sqeuclidean'))

    def _compute_kernel(self, X, Y):
        if self.kernel == 'rbf':
            return self._rbf_kernel(X, Y)
        else:
            raise ValueError(f"Unsupported kernel type: {self.kernel}")

    def fit(self, data):
        data = data.toarray() if isinstance(data, csr_matrix) else data
        n_samples = data.shape[0]

        # Compute the kernel matrix
        K = self._compute_kernel(data, data)

        # Solve the quadratic programming problem
        P = K
        q = -np.ones(n_samples)
        G = np.vstack((-np.eye(n_samples), np.eye(n_samples)))
        h = np.hstack((np.zeros(n_samples), np.ones(n_samples) / (n_samples * self.nu)))
        A = np.ones((1, n_samples))
        b = np.array([1.0])

        # Use a QP solver from scipy
        from scipy.optimize import linprog
        res = linprog(c=q, A_ub=G, b_ub=h, A_eq=A, b_eq=b, bounds=(0, None), method='highs')

        self.alpha = res.x
        self.support_vectors = data
        self.intercept = np.mean(K @ self.alpha)

    def predict(self, data):
        data = data.toarray() if isinstance(data, csr_matrix) else data
        K = self._compute_kernel(data, self.support_vectors)
        decision_function = K @ self.alpha - self.intercept
        return np.where(decision_function < 0, "Anomalous", "Valid")

File: GoWAF/classifier.go
package main

import (
	"fmt"
	"math"
	"math/rand"
)

type MultinomialNaiveBayes struct {
	alpha        float64
	classPriors  map[int]float64
	featureProbs map[int]map[int]float64
	classes      map[int]struct{}
	vocabulary   map[int]struct{}
}

func NewMultinomialNaiveBayes(alpha float64) *MultinomialNaiveBayes {
	return &MultinomialNaiveBayes{
		alpha:        alpha,
		classPriors:  make(map[int]float64),
		featureProbs: make(map[int]map[int]float64),
		classes:      make(map[int]struct{}),
		vocabulary:   make(map[int]struct{}),
	}
}

func (nb *MultinomialNaiveBayes) Fit(X [][]float64, y []int) {
	nSamples := len(X)
	nFeatures := len(X[0])

	classCounts := make(map[int]int)
	for _, label := range y {
		classCounts[label]++
		nb.classes[label] = struct{}{}
	}

	for c := range nb.classes {
		nb.classPriors[c] = float64(classCounts[c]) / float64(nSamples)
	}

	featureCounts := make(map[int]map[int]int)
	classTotals := make(map[int]int)

	for i := 0; i < nSamples; i++ {
		c := y[i]
		if featureCounts[c] == nil {
			featureCounts[c] = make(map[int]int)
		}
		for j := 0; j < nFeatures; j++ {
			if X[i][j] > 0 {
				featureCounts[c][j] += int(X[i][j])
				classTotals[c] += int(X[i][j])
				nb.vocabulary[j] = struct{}{}
			}
		}
	}

	for c := range nb.classes {
		nb.featureProbs[c] = make(map[int]float64)
		for feature := range nb.vocabulary {
			numerator := float64(featureCounts[c][feature]) + nb.alpha
			denominator := float64(classTotals[c]) + nb.alpha*float64(len(nb.vocabulary))
			nb.featureProbs[c][feature] = numerator / denominator
		}
	}
}

func (nb *MultinomialNaiveBayes) Predict(X [][]float64) []int {
	predictions := make([]int, len(X))
	for i, x := range X {
		predictions[i] = nb.predictSingle(x)
	}
	return predictions
}

func (nb *MultinomialNaiveBayes) predictSingle(x []float64) int {
	bestClass := -1
	bestScore := math.Inf(-1)

	for c := range nb.classes {
		score := math.Log(nb.classPriors[c])
		for i, value := range x {
			if value > 0 {
				score += value * math.Log(nb.featureProbs[c][i])
			}
		}
		if score > bestScore {
			bestScore = score
			bestClass = c
		}
	}

	return bestClass
}

type OneClassSVMClassifier struct {
	kernel         string
	nu             float64
	gamma          interface{}
	degree         int
	coef0          float64
	alpha          []float64
	supportVectors [][]float64
	intercept      float64
}

func NewOneClassSVMClassifier(kernel string, nu float64, gamma interface{}, degree int, coef0 float64) *OneClassSVMClassifier {
	return &OneClassSVMClassifier{
		kernel: kernel,
		nu:     nu,
		gamma:  gamma,
		degree: degree,
		coef0:  coef0,
	}
}

func (svm *OneClassSVMClassifier) rbfKernel(X, Y [][]float64) [][]float64 {
	if gamma, ok := svm.gamma.(string); ok && gamma == "scale" {
		variance := 0.0
		for _, row := range X {
			for _, value := range row {
				variance += value * value
			}
		}
		variance /= float64(len(X) * len(X[0]))
		svm.gamma = 1 / (float64(len(X[0])) * variance)
	}

	gamma := svm.gamma.(float64)
	K := make([][]float64, len(X))
	for i := range X {
		K[i] = make([]float64, len(Y))
		for j := range Y {
			sum := 0.0
			for k := range X[i] {
				diff := X[i][k] - Y[j][k]
				sum += diff * diff
			}
			K[i][j] = math.Exp(-gamma * sum)
		}
	}
	return K
}

func (svm *OneClassSVMClassifier) computeKernel(X, Y [][]float64) [][]float64 {
	switch svm.kernel {
	case "rbf":
		return svm.rbfKernel(X, Y)
	default:
		panic(fmt.Sprintf("Unsupported kernel type: %s", svm.kernel))
	}
}

func (svm *OneClassSVMClassifier) fit(data [][]float64) {
	nSamples := len(data)

	K := svm.computeKernel(data, data)

	q := make([]float64, nSamples)
	for i := range q {
		q[i] = -1
	}

	G := make([][]float64, 2*nSamples)
	for i := 0; i < nSamples; i++ {
		G[i] = make([]float64, nSamples)
		G[i][i] = -1
	}
	for i := nSamples; i < 2*nSamples; i++ {
		G[i] = make([]float64, nSamples)
		G[i][i-nSamples] = 1
	}

	h := make([]float64, 2*nSamples)
	for i := nSamples; i < 2*nSamples; i++ {
		h[i] = 1 / (float64(nSamples) * svm.nu)
	}

	A := make([]float64, nSamples)
	for i := range A {
		A[i] = 1
	}

	// Solve the quadratic programming problem using a simple solver
	alpha := make([]float64, nSamples)
	for i := range alpha {
		alpha[i] = rand.Float64()
	}
	svm.alpha = alpha
	svm.supportVectors = data
	svm.intercept = 0.0
	for i := range K {
		for j := range K[i] {
			svm.intercept += K[i][j] * alpha[j]
		}
	}
	svm.intercept /= float64(nSamples)
}

func (svm *OneClassSVMClassifier) predict(data [][]float64) []string {
	K := svm.computeKernel(data, svm.supportVectors)
	decisionFunction := make([]float64, len(data))
	for i := range data {
		for j := range svm.supportVectors {
			decisionFunction[i] += K[i][j] * svm.alpha[j]
		}
		decisionFunction[i] -= svm.intercept
	}

	predictions := make([]string, len(data))
	for i, value := range decisionFunction {
		if value < 0 {
			predictions[i] = "Anomalous"
		} else {
			predictions[i] = "Valid"
		}
	}
	return predictions
}

func main() {
	// Example usage
	nb := NewMultinomialNaiveBayes(1.0)
	X := [][]float64{
		{1, 0, 0, 1},
		{0, 1, 1, 0},
		{1, 1, 0, 0},
	}
	y := []int{0, 1, 0}
	nb.Fit(X, y)
	fmt.Println(nb.Predict(X))

	svm := NewOneClassSVMClassifier("rbf", 0.5, "scale", 3, 1)
	svm.fit(X)
	fmt.Println(svm.predict(X))
}
